docs/proposals/custom-metrics-api.md

```diff
@@ -0,0 +1,189 @@
+Custom Metrics API
+==================
+
+The new [metrics monitoring vision]() proposes an API that the Horizontal
+Pod Autoscaler can use to access arbitrary metrics.
+
+Similarly to the [master metrics API](resource-metrics-api.md), the new
+API should be structured around accessing metrics by referring to
+kubernetes objects (or groups thereof) and a metric name.  For this
+reason, the API could be useful for other consumers (most likely
+controllers) that want to consume custom metrics (similarly to how the
+master metrics API is generally useful to multiple cluster components).
+
+The HPA can currently refer to all pods selected by a label selector, and
+is planned to be able to support referring to metrics describing
+a namespace and other non-pod objects within the namespace.
+
+API Paths
+---------
+
+The root API path will look like `/apis/general-metrics/v1alpha1`.  For
+brevity, this will be left off below.
+
+- `/nodes/{node-name}/metrics/{metric-name...}`: retrieve the given metric
+  on the given node
+
+- `/namespaces/{namespace-name}/metrics/{metric-name...}`: retrieve the
+  given metric on the given namespace
+
+- `/namespaces/{namespace-name}/object-metrics/{object-type}/{metric-name...}`:
+  retrieve the given metric for all objects of the given type in the given
+  namespace.
+
+- `/namespaces/{namespace-name}/object-metrics/{object-type}/{metric-name...}?labelSelector=foo`:
+  retrieve the given metric for all objects of the given type matching the
+  given label selector in the given namespace.
+
+- `/namespaces/{namespace-name}/object-metrics/{object-type}/{metric-name...}?names=foo,bar`:
+  retrieve the given metric for the objects of the given type with the
+  given name in the given namespace
+
+For example, to retrieve the custom metric "hits-per-second" for all pods
+matching "app=frontend` in the namespaces "webapp", the request might look
+like:
+
+`/apis/general-metrics/v1alpha1/namespaces/webapp/object-metrics/pods/hits-per-second?labelSelector=app%3Dfrontend`.
+
+API Objects
+-----------
+
+The request URLs listed above will return objects of the `Metrics`
+type, described below:
+
+```go
+
+// a list of values for a given metric for some set of objects
+type Metrics struct {
+    unversioned.TypeMeta `json:",inline"`
+    unversioned.ListMeta `json:"metadata,omitempty"`
+
+    // the name of the metric
+    MetricName string `json:"metricName"`
+
+    // the value of the metric across the described objects
+    MetricValues []MetricValue `json:"metricValues"`
+}
+
+// a metric value for some object
+type GenericMetrics struct {
```

Comment:
whoops, there's a typo above -- it should list `GenericMetrics` instead of `MetricsValue` above (to match the name scheme from the master metrics API -- we have NodeMetrics, PodMetrics, and GenericMetrics).